test/object-store/sync/flx_role_change.cpp

```diff
@@ -857,4 +857,299 @@ TEST_CASE("flx: role changes during bootstrap complete successfully", "[sync][fl
+                skip_role_change_check = skip_role_check;
+                return ClientResetTestState::start;
+            });
+        };
+
+        SECTION("Bind before client reset") {
+            logger->debug("ROLE CHANGE: Role change after BIND before client reset");
+            set_expected_role_state(ClientResetTestState::bind_before_cr_session, true);
+        }
+        SECTION("Client reset session ident") {
+            logger->debug("ROLE CHANGE: Role change after client reset session IDENT");
+            set_expected_role_state(ClientResetTestState::cr_session_ident);
+        }
+        SECTION("Client reset session downloading") {
+            logger->debug("ROLE CHANGE: Role change while client reset session downloading");
+            // skip the role change error check since it may not be received before the fresh
+            // realm download is complete
+            set_expected_role_state(ClientResetTestState::cr_session_downloading, true);
+        }
+        SECTION("Client reset session downloaded") {
+            logger->debug("ROLE CHANGE: Role change after client reset session donwloaded");
+            set_expected_role_state(ClientResetTestState::cr_session_downloaded);
+        }
+        SECTION("Client reset session integrating") {
+            logger->debug("ROLE CHANGE: Role change after client reset session integrating");
+            set_expected_role_state(ClientResetTestState::cr_session_integrating);
+        }
+        SECTION("Client reset session integrated") {
+            logger->debug("ROLE CHANGE: Role change after client reset session integrated");
+            set_expected_role_state(ClientResetTestState::cr_session_integrated);
+        }
+        SECTION("BIND after client reset session") {
+            logger->debug("ROLE CHANGE: Role change after BIND after client reset session");
+            set_expected_role_state(ClientResetTestState::bind_after_cr_session, true);
+        }
+        SECTION("Merged after client reset session") {
+            logger->debug("ROLE CHANGE: Role change after merge after client reset session");
+            set_expected_role_state(ClientResetTestState::merged_after_cr_session, true);
+        }
+        SECTION("Merged after client reset session") {
+            logger->debug("ROLE CHANGE: Role change after IDENT after client reset session");
+            set_expected_role_state(ClientResetTestState::ident_after_cr_session);
+        }
+
+        // Client reset will happen when session tries to reconnect
+        realm_1->sync_session()->restart_session();
+        auto resync_mode = wait_for_future(std::move(reset_future)).get();
+        wait_for_update(realm_1);
+
+        // Verify the data was downloaded/updated (only the employee records)
+        verify_records(realm_1, params.num_emps, 0, 0);
+
+        client_reset_state.transition_with([&](ClientResetTestState) {
```

Comment:
do we need this last transition?

Reply:
It's not really doing a transition, but reusing the state machine's mutex to guard these values between the event loop and the main test thread.